pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -70,29 +100,66 @@ func (e *HookExecutor) executeExecNewPod(hook *deployapi.LifecycleHook, deployme
 	}
 
 	// Try to create the pod.
-	pod, err := e.PodClient.CreatePod(deployment.Namespace, podSpec)
+	pod, err := e.podClient.CreatePod(deployment.Namespace, podSpec)
 	if err != nil {
 		if !kerrors.IsAlreadyExists(err) {
 			return fmt.Errorf("couldn't create lifecycle pod for %s: %v", deployutil.LabelForDeployment(deployment), err)
 		}
 	} else {
-		glog.V(0).Infof("Created lifecycle pod %s for deployment %s", pod.Name, deployutil.LabelForDeployment(deployment))
+		glog.V(0).Infof("Created lifecycle pod %s/%s for deployment %s", pod.Namespace, pod.Name, deployutil.LabelForDeployment(deployment))
 	}
 
 	stopChannel := make(chan struct{})
 	defer close(stopChannel)
-	nextPod := e.PodClient.PodWatch(pod.Namespace, pod.Name, pod.ResourceVersion, stopChannel)
-
-	glog.V(0).Infof("Waiting for hook pod %s/%s to complete", pod.Namespace, pod.Name)
+	nextPod := e.podClient.PodWatch(pod.Namespace, pod.Name, pod.ResourceVersion, stopChannel)
+
+	// Wait for the hook pod to reach a terminal phase. Start reading logs as
+	// soon as the pod enters a usable phase.
+	var updatedPod *kapi.Pod
+	var once sync.Once
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	glog.V(0).Infof("Watching logs for hook pod %s/%s while awaiting completion", pod.Namespace, pod.Name)
+waitLoop:
 	for {
-		pod := nextPod()
-		switch pod.Status.Phase {
-		case kapi.PodSucceeded:
-			return nil
-		case kapi.PodFailed:
-			return fmt.Errorf(pod.Status.Message)
+		updatedPod = nextPod()
+		switch updatedPod.Status.Phase {
+		case kapi.PodRunning:
+			go once.Do(func() { e.readPodLogs(pod, wg) })
+		case kapi.PodSucceeded, kapi.PodFailed:
+			go once.Do(func() { e.readPodLogs(pod, wg) })
+			break waitLoop
 		}
 	}
+	// The pod is finished, wait for all logs to be consumed before returning.
```

Comment:
We're not necessarily waiting for the logs to be consumed, we're waiting for our *attempt* to consume the logs to be completed. There's no use asking for the logs until the pod's in a phase where the server will return them, and even though the pod was observed to be in a log-readable phase, by the time we ask for the stream, the master might return an error or a broken stream anyway. We also need to block until the pod is terminal anyway: an end of the log stream doesn't necessarily mean the pod has reached a terminal phase. It doesn't seem good to ignore the phase updates while blocking synchronously on log streaming.

If you have another algorithm to propose, describe it here and we'll discuss.